File: app/src/main/java/com/moh/mhealth/Converters.java
package com.moh.mhealth;

import android.content.Intent;
import android.location.Location;

import androidx.room.TypeConverter;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Date;

public class Converters {

    @TypeConverter
    public static LocalDate toLocalDate(long epochDay) {
        return LocalDate.ofEpochDay(epochDay);
    }

    @TypeConverter
    public static long fromLocalDate(LocalDate date) {
        return date.toEpochDay();
    }

    @TypeConverter
    public static Date toDate(Long dateLong) {
        return dateLong == null ? null : new Date(dateLong);
    }

    @TypeConverter
    public static Long fromDate(Date date) {
        return date == null ? null : date.getTime();
    }

    @TypeConverter
    public static int[] toArray(String intStr) {
        String[] arr = intStr.split(",");

        int[] arrInt = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            arrInt[i] = Integer.parseInt(arr[i]);
        }

        return arrInt;
    }

    @TypeConverter
    public static String fromArray(int[] arrInt) {
        if (arrInt.length == 0) {
            return "";
        }
        String str = arrInt[0] + "";

        for (int i = 1; i < arrInt.length; i++) {
            str += ("," + arrInt[i]);
        }

        return str;
    }


    @TypeConverter
    public static String fromLocation(SimpleLocation loc) {
        return loc.getLatitude() + ";" + loc.getLongitude();
    }

    @TypeConverter
    public static SimpleLocation toLocation(String locStr) {
        String[] latLong = locStr.split(";");
        return new SimpleLocation(Double.parseDouble(latLong[0]), Double.parseDouble(latLong[1]));
    }
}

File: app/src/main/java/com/moh/mhealth/Converters.kt
package com.moh.mhealth

import androidx.room.TypeConverter
import java.time.LocalDate
import java.util.*

object Converters {
    @JvmStatic
    @TypeConverter
    fun toLocalDate(epochDay: Long): LocalDate {
        return LocalDate.ofEpochDay(epochDay)
    }

    @JvmStatic
    @TypeConverter
    fun fromLocalDate(date: LocalDate): Long {
        return date.toEpochDay()
    }

    @JvmStatic
    @TypeConverter
    fun toDate(dateLong: Long?): Date? {
        return if (dateLong == null) null else Date(dateLong)
    }

    @JvmStatic
    @TypeConverter
    fun fromDate(date: Date?): Long? {
        return date?.time
    }

    @JvmStatic
    @TypeConverter
    fun toArray(intStr: String): IntArray {
        val arr = intStr.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val arrInt = IntArray(arr.size)
        for (i in arr.indices) {
            arrInt[i] = arr[i].toInt()
        }
        return arrInt
    }

    @JvmStatic
    @TypeConverter
    fun fromArray(arrInt: IntArray): String {
        if (arrInt.size == 0) {
            return ""
        }
        var str = arrInt[0].toString() + ""
        for (i in 1 until arrInt.size) {
            str += "," + arrInt[i]
        }
        return str
    }

    @JvmStatic
    @TypeConverter
    fun fromLocation(loc: SimpleLocation?): String {
        return if (loc == null) {
            ""
        } else loc.latitude.toString() + ";" + loc.longitude
    }

    @JvmStatic
    @TypeConverter
    fun toLocation(locStr: String): SimpleLocation? {
        val latLong = locStr.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return if (locStr.length < 2) {
            null
        } else SimpleLocation(
            latLong[0].toDouble(),
            latLong[1].toDouble()
        )
    }
}